pkg/theiactl/portforwarder/portforwarder.go

```diff
@@ -0,0 +1,179 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package portforwarder
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+type PortForwarder struct {
+	Config        *rest.Config
+	Clientset     kubernetes.Interface
+	Namespace     string
+	Name          string
+	TargetPort    int
+	ListenAddress string
+	ListenPort    int
+	stopCh        chan struct{}
+}
+
+// This function creates Port Forwarder for a Pod
+func NewPortForwarder(config *rest.Config, namespace string, pod string, listenAddress string, listenPort int, targetPort int) (*PortForwarder, error) {
+
+	pf := &PortForwarder{
+		Config:        config,
+		Namespace:     namespace,
+		Name:          pod,
+		TargetPort:    targetPort,
+		ListenAddress: listenAddress,
+		ListenPort:    listenPort,
+	}
+
+	var err error
+	pf.Clientset, err = kubernetes.NewForConfig(pf.Config)
+	if err != nil {
+		return pf, fmt.Errorf("Could not create kubernetes client: %v", err)
+	}
+
+	return pf, nil
+}
+
+// This function creates Port Forwarder for a Service by finding first Pod
+// that belongs to the Service, and target Port for requested Service Port.
+// This code is based upon kubectl port-forward implementation
+func NewServicePortForwarder(config *rest.Config, namespace string, service string, listenAddress string, listenPort int, servicePort int) (*PortForwarder, error) {
+
+	pf := &PortForwarder{
+		Namespace:     namespace,
+		ListenAddress: listenAddress,
+		ListenPort:    listenPort,
+	}
+
+	var err error
+	pf.Clientset, err = kubernetes.NewForConfig(pf.Config)
+	if err != nil {
+		return pf, fmt.Errorf("Could not create kubernetes client: %v", err)
+	}
+
+	serviceObj, err := pf.Clientset.CoreV1().Services(pf.Namespace).Get(context.TODO(), service, metav1.GetOptions{})
+	if err != nil {
+		return pf, fmt.Errorf("Failed to read Service %s: %v", service, err)
+	}
+
+	// find container port that corresponds to requested service port
+	pf.TargetPort, err = getContainerPortByServicePort(serviceObj, servicePort)
+	if err != nil {
+		return pf, err
+	}
+
+	listOptions := metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("app=%s", serviceObj.Spec.Selector["app"]),
+	}
+
+	// for target Pod - take first Pod for the Service
+	pods, err := pf.Clientset.CoreV1().Pods(pf.Namespace).List(context.TODO(), listOptions)
+
+	if err != nil {
+		return pf, fmt.Errorf("Failed to read Pods for Service %s: %v", service, err)
+	}
+	if len(pods.Items) == 0 {
+		return pf, fmt.Errorf("No Pods found for Service %s: %v", service, err)
+	}
+
+	pod := pods.Items[0]
```

Comment:
An important aspect of kubectl port forward is that you always forward to a pod, is that correct?
I don't think you can forward to a service and let k8s kube proxy do the endpoint selection

Reply:
The API only allows forwarding to a pod, however kubectl allows to forward to a service, and behind the scenes it would choose first pod for the service (this logic is copied in this PR)

Reply:
Understood, thanks for the explanation